src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -132,11 +133,25 @@ def get_preference(
         # while we work on "proper" branch pruning techniques.
         delay_this = identifier == "setuptools"
 
+
+        # Prefer the causes of failures. In general this speeds up backtracking,
+        # on the assumption that the problem resolving the dependency tree is
+        # related to the failures that are currently happening
+        responsible_for_failure = False
+        for failure_cause in failure_causes:
+            if identifier == failure_cause.requirement.name:
+                responsible_for_failure = True
+                break
+            if failure_cause.parent and identifier == failure_cause.parent.name:
+                responsible_for_failure = True
+                break
+
         return (
             not requires_python,
             delay_this,
             not direct,
             not pinned,
+            not responsible_for_failure,
```

Comment:
Hmm... This should likely be the first item in this tuple. We want to actively try variants of items that are involved in a conflict.

Reply:
You are likely correct. My reasoning was python, direct, and pinned should all be singular package requirements, and I didn't want to mess with pinning them first. But as "responsible_for_failures" only becomes relevant once there are failures I guess this makes sense. I will do some sanity testing though that this doesn't cause problems.

As for "delay_this" I actually personally feel that this setuptools hack can be removed with this change, as it won't be responsible for failures and therefore pip won't be "accidentally" backtracking on setuptools anymore. I just wasn't brave enough to do it on this PR.

Reply:
We can have a `--use-deprecated=setuptools-resolve-hack` that enables `delay_this` and point people report to a tracking issue when they use this so we can get an idea if it’s actually useful after this. But I agree this should be a separate PR.

Meanwhile, I would put `not responsible_for_failure` after `delay_this` (`requires_python` should always be the first one to make unresolvable Python version conflits bubble up as soon as possible).

Reply:
I am running through my known test cases and I found a requirements file where the proposed change here makes it significantly slower to resolve (in fact it hasn't finished resolving):

```
# Python 3.7 only
poetry2setup
wheel
twine
```

Without the proposed change it significantly backtracks on "keyring" to "19.2.0", then after 1 more significant backtrack on "importlib_metadata" it resolves quickly.

With the proposed change the resolver backtracks  on "keyring" to "15.1.0", backtracks significantly on "importlib_metadata" (but much slower), then backtracks significantly on "packaging", and that is as far as I've been able to get after an hour of backtracking.

I have not yet been able to figure out all the specific details. But I hypothesise it is very efficient to quickly check single requirements (e.g. direct and pinned), even versus following backtrack failures.

Let me know what you think.